Fixes impl self methods with where clause.

esdrubal opened the pull request:
When where clauses were used in impl self methods with a generic parameter from impl<T> an error was thrown saying `cannot find type "T" in this scope`

This fix includes multiple changes:
  - Parent type parameters that are used in method where clauses are now added to `FunctionDeclaration.type_parameters`
  - `type_check_method_application` now replaces decls from trait constraints.
  - `insert_trait_implementation` and `get_methods_for_type_and_trait_name` now use a new method `trait_name.to_fullpath` this improves consistency between inserting and getting by using a full path representation in stored trait implementations.
  - `Namespace::init_root` now receives a pkg_name, this was required by the previous change.

Closes #3333

Another issue this solves is related to #3325 where multiple trait constraints were used for same method but with different types an error would occur as the replace of the declaration would pick only one trait method implementation for both constraints.

The solution was to use IdentUnique while collecting traits implemented methods in `gather_decl_mapping_from_trait_constraints`. This allows to create `DeclMapping` mappings from one trait original method to multiple trait implementations. For instance eq method original method to both bool and u64 trait implementations.

The other step of the solution was to verify while replacing decls if the current function application beeing replaced matches the argument types of one decl mapping specifically.

This changes also fix #3623, added test case `generic_result_method`.
Closes #3623

mohammadfawaz commented:
Thank you @esdrubal. Nice fix! Before we start reviewing this, do you mind adding a lot more tests that cover a variety of different situations such as:
- multiple trait constraints.
- multiple methods in one impl block with different multiple trait constraints in them
- multiple impl blocks

And so on. The more complex the tests the better. I don't believe this is a feature that we have any tests for at the moment, so it would be great to have a whole list of tests for full coverage.

esdrubal commented:
I will add more tests.

emilyaherbert commented on test/src/e2e_vm_tests/test_programs/should_pass/language/generic_impl_self_where/src/main.sw:
```diff
@@ -0,0 +1,81 @@
+
+  assert(d.contains6(Data2 { x: 42, y: 42 }));
+  assert(!d.contains6(Data2 { x: 42, y: 41 }));
+  assert(!d.contains6(Data2 { x: 41, y: 42 }));
```
Can you expand this example to test for a few more types in addition to `numeric`?

emilyaherbert commented on sway-core/src/type_system/type_parameter.rs:
```diff
@@ -22,6 +22,7 @@ pub struct TypeParameter {
     pub name_ident: Ident,
     pub(crate) trait_constraints: Vec<TraitConstraint>,
     pub(crate) trait_constraints_span: Span,
+    pub(crate) is_from_parent: bool,
```
I don't know if this field is ever set to `true` anywhere ?

emilyaherbert commented on sway-core/src/type_system/type_parameter.rs:
```diff
@@ -137,22 +139,50 @@ impl TypeParameter {
-        ctx.namespace
-            .insert_symbol(name_ident.clone(), type_parameter_decl)
-            .ok(&mut warnings, &mut errors);
+        if is_from_parent {
```
This function is only called when type parameters are declarated. i.e. when we see `fn my_function<T, E>(..)` and `impl<V> Data<V> { .. }`, is that what you are intending? Otherwise, the "else" path that you have here is how the type parameters are distributed to the other instances in which those type parameters are used.

emilyaherbert commented on sway-core/src/language/ty/expression/expression_variant.rs:
```diff
@@ -614,7 +617,46 @@ impl ReplaceDecls for TyExpressionVariant {
-                function_decl_id.replace_decls(decl_mapping, type_engine);
+                // Filter decl mapping so function declaration is replaced by another
+                // function declaration that matches argument types
+                // This is needed for when multiple trait constraints are used that replace the same trait function
```
Can you elaborate on this? Maybe a code example? Thanks!

esdrubal replied:
This is required when we have a function such as:
```rust
fn function<T, E>(v1: T, v2: E) where T: Eq, E: Eq {
  v1 == v1
  v2 == v2
}
```
if T and E are different types then `v1 == v1` should call one trait implementation while `v2 == v2` should call another one. In this case decl_mapping will have `v1 == v1` function declarations replaced by the T implementation and  `v2 == v2` function declarations replaced by the E implementation. The code below filters the correct implementation to use.

emilyaherbert commented on sway-core/src/type_system/type_parameter.rs:
```diff
@@ -22,6 +22,7 @@ pub struct TypeParameter {
     pub name_ident: Ident,
     pub(crate) trait_constraints: Vec<TraitConstraint>,
     pub(crate) trait_constraints_span: Span,
+    pub(crate) is_from_parent: bool,
```
Ah nevermind, I see that it is set to true during parse time---we should leave analysis until the semantic analysis stage.

emilyaherbert reviewed (changes requested):
High level review comments:
- One of the changes added in this PR is to detect at parse time which type parameters are from a parent. We should leave that analysis to be done during semantic analysis, not at parse time.
- One of the comments in the PR description says that "Parent type parameters that are used in method where clauses are now added to FunctionDeclaration.type_parameters". This already is in the compiler prior to this PR, see: https://github.com/FuelLabs/sway/blob/0f1ca16cec81a3e6a866fc3c099b889519854598/sway-core/src/semantic_analysis/ast_node/declaration/impl_trait.rs#L845
- Should we consider using `CallPath`'s instead of `IdentUnique`'s?

esdrubal commented:
> * One of the changes added in this PR is to detect at parse time which type parameters are from a parent. We should leave that analysis to be done during semantic analysis, not at parse time.
> * One of the comments in the PR description says that "Parent type parameters that are used in method where clauses are now added to FunctionDeclaration.type_parameters". This already is in the compiler prior to this PR, see:

I added it in parse time mainly because [`generic_params_opt_to_type_parameters`](https://github.com/FuelLabs/sway/pull/3621/files#diff-dd8f23cc5406f8b442080cbce96df8616c62d1a4ec57ce2a37d6125651edf09aL755) is where type parameter are created and where trait constraints are added to the type parameters. This used in [`item_fn_to_function_declaration`](https://github.com/FuelLabs/sway/blob/b379872b64de4c4c6ebf28760262d9dc0ccb311d/sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs#L428). We need to create a type parameter from a parent with the constraint that is in the function so this looked the best place to do it, do you think it can be done elsewhere?

esdrubal commented on sway-core/src/type_system/type_parameter.rs:
```diff
@@ -137,22 +139,50 @@ impl TypeParameter {
-        ctx.namespace
-            .insert_symbol(name_ident.clone(), type_parameter_decl)
-            .ok(&mut warnings, &mut errors);
+        if is_from_parent {
```
The else part is what we did with every type parameter before the changes. Type parameters with `is_from_parent` don't need the else part but still have to somehow make the association between the already defined generic type in the `impl<T>` and the generic type in the type parameter.

esdrubal commented on test/src/e2e_vm_tests/test_programs/should_pass/language/generic_impl_self_where/src/main.sw:
```diff
@@ -0,0 +1,81 @@
+
+  assert(d.contains6(Data2 { x: 42, y: 42 }));
+  assert(!d.contains6(Data2 { x: 42, y: 41 }));
+  assert(!d.contains6(Data2 { x: 41, y: 42 }));
```
Yes, I will add more tests that use different types.

esdrubal commented on sway-core/src/semantic_analysis/namespace/root.rs:
```diff
@@ -16,6 +16,7 @@ use super::{module::Module, namespace::Namespace, Path};
 #[derive(Clone, Debug)]
 pub struct Root {
     pub(crate) module: Module,
+    pub(crate) pkg_name: Option<Ident>,
```
Right now the content of `root.pkg_name` is the same of `root.module.name`, the only difference is that pkg_name is an Ident and module.name is a String. and pkg_name requires a leak such as `root.pkg_name = Some(Ident::new_no_span(Box::leak(pkg_name.into_boxed_str())))` to be created, which should only be one leak per compilation. To avoid more leaks we have this duplicated, but if we deem that a leak per module is acceptable we could remove `pkg_name` and use only `module.name`.

esdrubal commented on sway-core/src/semantic_analysis/namespace/namespace.rs:
```diff
@@ -286,6 +290,46 @@ impl Namespace {
     }
+
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn insert_trait_implementation(
```
`insert_trait_implementation` and  `get_methods_for_type_and_trait_name` now use `trait_name.to_fullpath(self)` which requires the self to be a `Namespace` instead of `Items` so they were moved to `Namespace`.

emilyaherbert replied:
SGTM

esdrubal commented on sway-core/src/semantic_analysis/namespace/module.rs:
```diff
@@ -40,6 +40,8 @@ pub struct Module {
     pub(crate) submodules: im::OrdMap<ModuleName, Module>,
     /// The set of symbols, implementations, synonyms and aliases present within this module.
     items: Items,
+    /// Name of the module, package name for root module, dependency name for other modules
```
`module.name` from a dependency comes from [here](https://github.com/esdrubal/sway/blob/b379872b64de4c4c6ebf28760262d9dc0ccb311d/sway-core/src/semantic_analysis/namespace/module.rs#L150) which is called with dep_name in [here](https://github.com/esdrubal/sway/blob/b379872b64de4c4c6ebf28760262d9dc0ccb311d/forc-pkg/src/pkg.rs#L2210).

emilyaherbert commented:
> We need to create a type parameter from a parent with the constraint that is in the function

Could you clarify how the approach in this PR is different from the code right [here](https://github.com/FuelLabs/sway/blob/0f1ca16cec81a3e6a866fc3c099b889519854598/sway-core/src/semantic_analysis/ast_node/declaration/impl_trait.rs#L845)? This bit of code takes the type parameters that are (1) present in the parent and (2) not present in the function.

emilyaherbert commented:
I like that this PR approaches this problem from a fresh perspective, but I feel like this approach of adding `is_from_parent` is moving too much of the semantic analysis work into parse time.

I believe that you've correctly identified that the bug is in translating type parameters from parents. But considering that in the original code example:
```rust
struct Vec<T> {
    x: T
}

impl<T> Vec<T> {
    fn contains(self, other: T) -> bool where T: Eq { 
        true
    }
}
```
```bash
10 | impl<T> Vec<T> {
11 |     fn contains(self, other: T) -> bool where T: Eq {
   |                                               ^ cannot find type "T" in this scope
12 |         true
13 |     }
   |
```
there is only one error message on the trait constraint and no error message on the argument `other: T`, this is giving me a hunch that this is a bug with how trait constraints _receive_/_view_ type parameters instead of how parents _push_ type parameters.

How about this---it seems that this PR fixes a number of issues, do you think that any of these issues could be fixed if you separated a subset of the relevant code into a separate PR? Maybe #3623 first? I'm wondering if there are a couple of different issues intermingling, and maybe chopping them off one at a time will give us a better picture?

esdrubal commented:
> Could you clarify how the approach in this PR is different from the code right [here](https://github.com/FuelLabs/sway/blob/0f1ca16cec81a3e6a866fc3c099b889519854598/sway-core/src/semantic_analysis/ast_node/declaration/impl_trait.rs#L845)?

You are right with this changes we have type parameters repeated from the parent, but this solution still depends on adding parent generic types in `generic_params_opt_to_type_parameters` so we can store the constraint along with it. I think one option we have is to move the code you mentioned to `generic_params_opt_to_type_parameters`. Another option would be to have another field and data structure in FunctionDeclaration that would store trait constraints to be applied during the semantical analysis.

> How about this---it seems that this PR fixes a number of issues, do you think that any of these issues could be fixed if you separated a subset of the relevant code into a separate PR?

Ok, I will make other PRs with subsets of this changes.

emilyaherbert commented:
@esdrubal, I gave this PR more thought and I've come up with a plan outline for how I propose that you proceed.

The error message we are seeing is [this variant](https://github.com/FuelLabs/sway/blob/0a3beae92e6faed5e8ca35da5a8071e58f4a43d1/sway-error/src/convert_parse_tree_error.rs#L75), and is produced [here](https://github.com/FuelLabs/sway/blob/0a3beae92e6faed5e8ca35da5a8071e58f4a43d1/sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs#L921). This is because this lookup into `params` is unable to find the type parameters of the parents. Instead of tracking which type parameters are from the parents etc (the previous approach on this PR), I propose that we remove the [`trait_constraints` field on `TypeParameter`](https://github.com/FuelLabs/sway/blob/0a3beae92e6faed5e8ca35da5a8071e58f4a43d1/sway-core/src/type_system/type_parameter.rs#L26) and add a new `trait_constraints` field onto [`FunctionDeclaration`](https://github.com/FuelLabs/sway/blob/0a3beae92e6faed5e8ca35da5a8071e58f4a43d1/sway-core/src/language/parsed/declaration/function.rs#L10) (and the other data structures that allow for where clauses as well).

The benefits that this approach offers us are:
1. we can implement this right now without waiting for other PR's to go in
2. reduces new abstraction overhead over other approaches
3. it is fairly (arguably _more_) ergonomic to place the trait constraints on `FunctionDeclaration` in the AST rather than in the `TypeParameters`. And also I believe a dive into the codebase would show that the `trait_constraints` field on `TypeParameters` versus `FunctionDeclaration` is not a strict requirement for correct compiler function.

What do you think @esdrubal? CC @mohammadfawaz

esdrubal commented:
> 3\. it is fairly (arguably _more_) ergonomic to place the trait constraints on `FunctionDeclaration` in the AST rather than in the `TypeParameters`. And also I believe a dive into the codebase would show that the `trait_constraints` field on `TypeParameters` versus `FunctionDeclaration` is not a strict requirement for correct compiler function.

@emilyaherbert yes I also think it would be better to have `trait_constraints` directly on `FunctionDeclaration`. I will work on a PR with that change.
